### ISSUE TITLE
IAsyncDisposable Singleton causes exception during aspnet.core application shutdown

### ISSUE THREAD

**Issue by gkello** [simpleinjector/SimpleInjector#872]
## Describe the bug

If a singleton implementing only IAsyncDisposable interface is registered in aspnet.core application, an exception is thrown during application shutdown

## Expected behavior
Application shuts down cleanly

## Actual behavior
```
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
Unhandled exception. System.InvalidOperationException: Singleton only implements IAsyncDisposable, but not IDisposable. Make sure to call Scope.DisposeAsync() instead of Dispose().
   at SimpleInjector.Scope.DisposeInstancesInReverseOrder(List`1 disposables, Int32 startingAsIndex)
   at SimpleInjector.Scope.DisposeRecursively(Boolean operatingInException)
   at SimpleInjector.Scope.Dispose(Boolean disposing)
   at SimpleInjector.Scope.Dispose()
   at SimpleInjector.ContainerScope.Dispose()
   at SimpleInjector.Container.Dispose(Boolean disposing)
   at SimpleInjector.Container.Dispose()
   at SimpleInjector.SimpleInjectorServiceCollectionExtensions.ContainerDisposeWrapper.Dispose()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.DisposeAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Extensions.Hosting.Internal.Host.DisposeAsync()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at SimpleInjectorTest.Program.Main(String[] args) in C:\temp\SimpleInjectorTest\Program.cs:line 16
```

## To Reproduce

1. Create new asp.net core project (netcore 3.1)
2. Add simpleinjector packages
```
    <PackageReference Include="SimpleInjector" Version="5.1.0" />
    <PackageReference Include="SimpleInjector.Integration.AspNetCore" Version="5.1.0" />
```
3. Add simpleinjector configuration code to startup class, register singleton
4. compile and run application in console. (Change startup type in visual studio)
5. Close the application (ctrl+C)
6. Exception occurs

## Additional context

Discovered by using Masstransit.SimpleInjector integration, https://github.com/MassTransit/MassTransit/issues/2139

Startup.cs file
``` c#
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using SimpleInjector;
using System;
using System.Threading.Tasks;

namespace SimpleInjectorTest
{
    public class Singleton : IAsyncDisposable
    {
        public async ValueTask DisposeAsync()
        {
            await Task.CompletedTask;
        }
    }

    public class Startup
    {
        Container container = new Container();

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddSimpleInjector(container);
            container.RegisterSingleton<Singleton>();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            app.UseSimpleInjector(container);
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            app.UseRouting();
            app.UseEndpoints(endpoints =>
            {
                endpoints.MapGet("/", async context =>
                {
                    await context.Response.WriteAsync("Hello World!");
                });
            });

            container.Verify();
        }
    }
}
```

**Comment by dotnetjunkie** [comment]
Are you using the latest version of `SimpleInjector.Integration.ServiceCollection`? The latest version is v5.0.2.

**Comment by gkello** [comment]
No I was not. Adding it fixes the issue.

Seems like SimpleInjector.Integration.AspNetCore should have dependency version increased, currently refers SimpleInjector.Integration.ServiceCollection v5.0.0

**Comment by dotnetjunkie** [comment]
This bug was fixed in [v5.0.2](https://github.com/simpleinjector/SimpleInjector.Integration.AspNetCore/releases/tag/sc_5.0.2).

> Seems like SimpleInjector.Integration.AspNetCore should have dependency version increased, currently (v5.1.0) refers SimpleInjector.Integration.ServiceCollection v5.0.0

This is a strategy we deliberately took with v5. In v5, the strategy is to only update a package when its own DLLs have updates. In this case, the bug fix is in SimpleInjector.Integration.ServiceCollection and not in SimpleInjector.Integration.AspNetCore, hense the release of SimpleInjector.Integration.ServiceCollection and not SimpleInjector.Integration.AspNetCore. SimpleInjector.Integration.AspNetCore requires SimpleInjector.Integration.ServiceCollection 5.0.0 or higher. That's still the case, even with the bug fix.

Reason for choosing this strategy is because it would otherwise cause a cascading effect through all dependent packages (which we say in v3 and v4) where we released new versions of NuGet packages that (themselves) didn't have any changes. This is confusing.

On the other hand, we now have 6 months experience with this new model, and it isn't working really well either. That's mainly because NuGet **does not show you updates for packages that were implicitly installed by NuGet because they are dependenies of packages that the developer explicitly installed**. This really, really sucks. It means that the developer has to traverse the dependency chain and install everything explicitly. This is hell.

I will think about this what the best solution is going forward.

**Comment by dotnetjunkie** [comment]
@gkello, I pushed an 5.2.0-alpha2 version of the Simple Injector core library and the ASP.NET Core integration packages to NuGet. These set of packages tend to solve the problems regarding `Microsoft.bcl.AsyncInterfaces` completely. Would you mind trying these packages out to see if this solves the problem for you and whether there are any unintended consequences? You can see more details about what changes here: #867. This is also the place to post your feedback. Thanks in advance.
